File: java/03-bootstrap/src/main/java/com/fResult/bootstrap/customers/services/TransactionTemplateCustomerService.java
package com.fResult.bootstrap.customers.services;

import com.fResult.bootstrap.customers.Customer;
import java.util.Collection;
import javax.sql.DataSource;
import org.springframework.transaction.support.TransactionTemplate;

public class TransactionTemplateCustomerService extends BaseCustomerService {
  private final TransactionTemplate transactionTemplate;

  public TransactionTemplateCustomerService(
      DataSource dataSource, TransactionTemplate transactionTemplate) {

    super(dataSource);
    this.transactionTemplate = transactionTemplate;
  }

  @Override
  public Collection<Customer> save(String... names) {
    return transactionTemplate.execute(ignored -> super.save(names));
  }

  @Override
  public Customer findById(Long id) {
    return transactionTemplate.execute(ignored -> super.findById(id));
  }

  @Override
  public Collection<Customer> findAll() {
    return transactionTemplate.execute(ignored -> super.findAll());
  }
}

File: kotlin/03-bootstrap/src/main/kotlin/com/fResult/bootstrap/customers/services/TransactionTemplateCustomerService.kt
package com.fResult.bootstrap.customers.services

import com.fResult.bootstrap.customers.Customer
import org.springframework.transaction.support.TransactionTemplate
import javax.sql.DataSource

open class TransactionTemplateCustomerService(
  dataSource: DataSource,
  private val transactionTemplate: TransactionTemplate,
) : BaseCustomerService(dataSource) {

  override fun save(vararg names: String): Collection<Customer> =
    transactionTemplate.execute { super.save(*names) } ?: listOf()

  override fun findById(id: Long): Customer? = transactionTemplate.execute { super.findById(id) }

  override fun findAll(): Collection<Customer> = transactionTemplate.execute { super.findAll() } ?: listOf()
}
